File: src/parser/lookups.go
package parser

import (
	"github.com/tlaceby/bedrock/src/ast"
	"github.com/tlaceby/bedrock/src/lexer"
)

type binding_power int

const (
	defalt_bp binding_power = iota
	comma
	assignment
	logical
	relational
	additive
	multiplicative
	unary
	call
	member
	primary
)

type stmt_handler func(p *parser) ast.Stmt
type nud_handler func(p *parser) ast.Expr
type led_handler func(p *parser, left ast.Expr, bp binding_power) ast.Expr

type stmt_lookup map[lexer.TokenKind]stmt_handler
type nud_lookup map[lexer.TokenKind]nud_handler
type led_lookup map[lexer.TokenKind]led_handler
type bp_lookup map[lexer.TokenKind]binding_power

var bp_lu = bp_lookup{}
var nud_lu = nud_lookup{}
var led_lu = led_lookup{}
var stmt_lu = stmt_lookup{}

func led(kind lexer.TokenKind, bp binding_power, led_fn led_handler) {
	bp_lu[kind] = bp
	led_lu[kind] = led_fn
}

func nud(kind lexer.TokenKind, nud_fn nud_handler) {
	nud_lu[kind] = nud_fn
}

func stmt(kind lexer.TokenKind, stmt_fn stmt_handler) {
	stmt_lu[kind] = stmt_fn
}

func createTokenLookups() {
	// Assignment
	led(lexer.ASSIGNMENT, assignment, parse_assignment_expr)
	led(lexer.PLUS_EQUALS, assignment, parse_assignment_expr)
	led(lexer.MINUS_EQUALS, assignment, parse_assignment_expr)
	led(lexer.SLASH_EQUALS, assignment, parse_assignment_expr)
	led(lexer.STAR_EQUALS, assignment, parse_assignment_expr)

	// Logical
	led(lexer.AND, logical, parse_binary_expr)
	led(lexer.OR, logical, parse_binary_expr)
	led(lexer.DOT_DOT, logical, parse_range_expr)

	// Relational
	led(lexer.LESS, relational, parse_binary_expr)
	led(lexer.LESS_EQUALS, relational, parse_binary_expr)
	led(lexer.GREATER, relational, parse_binary_expr)
	led(lexer.GREATER_EQUALS, relational, parse_binary_expr)
	led(lexer.EQUALS, relational, parse_binary_expr)
	led(lexer.NOT_EQUALS, relational, parse_binary_expr)

	// Additive & Multiplicitave
	led(lexer.PLUS, additive, parse_binary_expr)
	led(lexer.DASH, additive, parse_binary_expr)
	led(lexer.SLASH, multiplicative, parse_binary_expr)
	led(lexer.STAR, multiplicative, parse_binary_expr)
	led(lexer.PERCENT, multiplicative, parse_binary_expr)

	// Literals & Symbols
	nud(lexer.NUMBER, parse_primary_expr)
	nud(lexer.STRING, parse_primary_expr)
	nud(lexer.IDENTIFIER, parse_primary_expr)

	// Unary/Prefix
	nud(lexer.TYPEOF, parse_prefix_expr)
	nud(lexer.DASH, parse_prefix_expr)
	nud(lexer.NOT, parse_prefix_expr)
	nud(lexer.OPEN_BRACKET, parse_array_literal_expr)

	// Member / Computed / Call
	led(lexer.DOT, member, parse_member_expr)
	led(lexer.COLON_COLON, member, parse_static_member_expr)
	led(lexer.OPEN_BRACKET, member, parse_member_expr)
	led(lexer.OPEN_PAREN, call, parse_call_expr)
	led(lexer.OPEN_CURLY, call, parse_struct_instantiation)
	led(lexer.OPEN_GENERIC, call, parse_generic_list_instantiation)

	// Grouping Expr
	nud(lexer.OPEN_PAREN, parse_grouping_expr)
	nud(lexer.FN, parse_fn_expr)

	// Macros
	nud(lexer.ASSERT_MACRO, parse_assert_macro)
	nud(lexer.STRING_MACRO, parse_string_macro)

	stmt(lexer.OPEN_CURLY, parse_block_stmt)
	stmt(lexer.LET, parse_var_decl_stmt)
	stmt(lexer.CONST, parse_var_decl_stmt)
	stmt(lexer.FN, parse_fn_declaration)
	stmt(lexer.IF, parse_if_stmt)
	stmt(lexer.IMPORT, parse_import_stmt)
	stmt(lexer.FOR, parse_for_stmt)
	stmt(lexer.WHILE, parse_while_stmt)
	stmt(lexer.STRUCT, parse_struct_declaration_stmt)
	stmt(lexer.RETURN, parse_return_stmt)
	stmt(lexer.TRAIT, parse_trait_stmt)
	stmt(lexer.MATCH, parse_match_stmt)
	stmt(lexer.UNSAFE, parse_unsafe_stmt)
	stmt(lexer.TYPEDEF, parse_typedef_stmt)
}

File: src/parser/lookups.cpp
#include "lookup.h"
#include "parser.h"

using namespace parser;
using namespace ast;
using namespace lexer;

std::unordered_map<lexer::TokenKind, type_nud_handler> parser::type_nud_lu{};
std::unordered_map<lexer::TokenKind, type_led_handler> parser::type_led_lu{};
std::unordered_map<lexer::TokenKind, BindingPower> parser::bp_lu{};
std::unordered_map<lexer::TokenKind, nud_handler> parser::nud_lu{};
std::unordered_map<lexer::TokenKind, BindingPower> parser::type_bp_lu{};
std::unordered_map<lexer::TokenKind, led_handler> parser::led_lu{};
std::unordered_map<lexer::TokenKind, stmt_handler> parser::stmt_lu{};

void parser::def_type_nud(TokenKind k, type_nud_handler h) {
  type_nud_lu[k] = h; //
}

void parser::def_type_led(TokenKind k, type_led_handler h, BindingPower bp) {
  type_bp_lu[k] = bp;
  type_led_lu[k] = h;
}

void parser::def_stmt(TokenKind kind, stmt_handler handler) {
  stmt_lu[kind] = handler; //
}

void parser::def_led(TokenKind kind, led_handler handler, BindingPower bp) {
  led_lu[kind] = handler;
  bp_lu[kind] = bp;
}

void parser::def_nud(TokenKind kind, nud_handler handler) {
  nud_lu[kind] = handler; //
}

inline bool parser::PRATT_PARSER_SETUP = false;
void parser::setup_pratt_parser() {
  if (PRATT_PARSER_SETUP)
    return;
  PRATT_PARSER_SETUP = true;
  setup_type_pratt_parser();

  // Stmts
  def_stmt(lexer::STRUCT, parse_struct_stmt);
  def_stmt(lexer::LET, parse_var_decl_stmt);
  def_stmt(lexer::CONST, parse_var_decl_stmt);
  def_stmt(lexer::FN, parse_fn_decl_stmt);
  def_stmt(lexer::OPEN_CURLY, parse_block_stmt);
  def_stmt(lexer::DEFER, parse_defer_stmt);
  def_stmt(lexer::IMPL, parse_impl_stmt);
  def_stmt(lexer::RETURN, parse_return_stmt);

  // NUD HANDLERS
  def_nud(lexer::OPEN_PAREN, parse_grouping_expr);
  def_nud(lexer::MINUS, parse_prefix_expr);
  def_nud(lexer::NOT, parse_prefix_expr);
  def_nud(lexer::STAR, parse_prefix_expr);
  def_nud(lexer::AMPERSAND, parse_prefix_expr);
  def_nud(lexer::NUMBER, parse_primary_expr);
  def_nud(lexer::STRING, parse_primary_expr);
  def_nud(lexer::IDENTIFIER, parse_primary_expr);

  // MACROS
  def_nud(lexer::FMT_MACRO, parse_fmt_macro);
  def_nud(lexer::STRING_MACRO, parse_str_macro);
  def_nud(lexer::NUMBER_MACRO, parse_num_macro);
  def_nud(lexer::LOG_MACRO, parse_log_macro);

  // LED HANLDERS
  def_led(lexer::ASSIGNMENT, parse_assignment_expr, ASSIGNMENT_BP);
  def_led(lexer::OPEN_PAREN, parse_call_expr, CALL_BP);

  def_led(lexer::PLUS, parse_binary_expr, ADDITIVE_BP);
  def_led(lexer::MINUS, parse_binary_expr, ADDITIVE_BP);

  def_led(lexer::STAR, parse_binary_expr, MULTIPLICATIVE_BP);
  def_led(lexer::SLASH, parse_binary_expr, MULTIPLICATIVE_BP);
  def_led(lexer::PERCENT, parse_binary_expr, MULTIPLICATIVE_BP);
}

void parser::setup_type_pratt_parser() {
  def_type_nud(lexer::IDENTIFIER, parse_symbol_type);
  def_type_nud(lexer::AMPERSAND, parse_pointer_type);
  def_type_nud(lexer::FN, parse_fn_type);
}